File: agent-mesh-py/python/agent_mesh/amesh/bind.py
"""``amesh bind github`` — cross-sign the user key with a GitHub SSH key.

Mirror of ``agent-mesh-cli/src/bind.rs``. The SSH key parse +
ed25519 cross-signature happen in Rust via
:class:`agent_mesh.core.GitHubBinding`; this module is the I/O shell
around it.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Optional

from agent_mesh.core import GitHubBinding, UserKey


def default_ssh_key_path() -> Path:
    """``~/.ssh/id_ed25519``. Matches the Rust CLI default."""
    return Path.home() / ".ssh" / "id_ed25519"


def github(
    home: Path,
    ssh_key: Optional[Path],
    username: Optional[str],
) -> int:
    """Produce a :class:`GitHubBinding` and persist it as JSON.

    Mirrors the Rust binding command's output format byte-for-byte:
    the binding is written as ``user.github.sig`` (pretty-printed JSON
    via ``json.dumps(..., indent=2)``), and stdout reports the binding
    path, the optional username hint, and the first 16 hex chars of
    the SSH pubkey for visual confirmation.
    """
    user_key_path = home / "user.key"
    if not user_key_path.exists():
        print(
            f"error: load user key at {user_key_path} — "
            "run `amesh keygen` first",
            file=sys.stderr,
        )
        return 1
    user = UserKey.load(str(user_key_path))

    ssh_path = ssh_key or default_ssh_key_path()
    if not ssh_path.exists():
        print(
            f"error: read SSH key from {ssh_path}: file not found",
            file=sys.stderr,
        )
        return 1
    try:
        ssh_pem_bytes = ssh_path.read_bytes()
    except OSError as e:
        print(f"error: read SSH key from {ssh_path}: {e}", file=sys.stderr)
        return 1

    try:
        binding = GitHubBinding.sign(user.public(), ssh_pem_bytes, username)
    except Exception as e:  # noqa: BLE001 — surface any signing failure as a clean CLI error
        print(f"error: {e}", file=sys.stderr)
        return 1

    binding_path = home / "user.github.sig"
    binding_path.parent.mkdir(parents=True, exist_ok=True)
    binding_path.write_text(json.dumps(binding.to_json(), indent=2))

    print(f"github binding written to {binding_path}")
    if username:
        print(f"username hint: {username}")
    print(
        "ssh pubkey fingerprint (first 16 hex chars): "
        f"{binding.ssh_pubkey_hex()[:16]}"
    )
    return 0

File: agent-mesh-cli/src/bind.rs
//! `amesh bind github` — cross-sign the user key with a GitHub SSH
//! ed25519 key.

use agent_mesh_core::{GitHubBinding, UserKey};
use anyhow::{Context, Result};
use ssh_key::PrivateKey as SshPrivateKey;
use std::path::PathBuf;

/// Read the user key, read the SSH key, produce a [`GitHubBinding`],
/// and write it to `<home>/user.github.sig` as pretty-printed JSON.
pub fn github(home: PathBuf, ssh_key: Option<PathBuf>, username: Option<String>) -> Result<()> {
    let user_key_path = home.join("user.key");
    let user = UserKey::load(&user_key_path).with_context(|| {
        format!(
            "load user key at {} — run `amesh keygen` first",
            user_key_path.display()
        )
    })?;

    let ssh_path = ssh_key.unwrap_or_else(default_ssh_key_path);
    let ssh_pem = std::fs::read_to_string(&ssh_path)
        .with_context(|| format!("read SSH key from {}", ssh_path.display()))?;
    let ssh = SshPrivateKey::from_openssh(&ssh_pem)
        .context("parse SSH key (must be ed25519, unencrypted)")?;

    let binding = GitHubBinding::sign(&user.public(), &ssh, username.clone())?;

    let binding_path = home.join("user.github.sig");
    if let Some(parent) = binding_path.parent() {
        if !parent.as_os_str().is_empty() {
            std::fs::create_dir_all(parent)?;
        }
    }
    let json = serde_json::to_string_pretty(&binding)?;
    std::fs::write(&binding_path, json)?;

    println!("github binding written to {}", binding_path.display());
    if let Some(u) = username {
        println!("username hint: {u}");
    }
    println!(
        "ssh pubkey fingerprint (first 16 hex chars): {}",
        &hex::encode(binding.ssh_pubkey)[..16]
    );
    Ok(())
}

/// Default SSH key location, `~/.ssh/id_ed25519`. Falls back to a
/// relative path if `$HOME` is unset.
fn default_ssh_key_path() -> PathBuf {
    dirs::home_dir()
        .map(|h| h.join(".ssh/id_ed25519"))
        .unwrap_or_else(|| PathBuf::from("id_ed25519"))
}
